src/main/java/com/ibm/watson/modelmesh/payload/AsyncPayloadProcessor.java

```diff
@@ -65,14 +74,42 @@ public AsyncPayloadProcessor(PayloadProcessor delegate, int delay, TimeUnit time
         }, 0, delay, timeUnit);
     }
 
+    void processPayload(Payload p) {
+        boolean released = false;
+        try {
+            released = delegate.process(p);
+        } catch (Throwable t) {
+            logger.warn("Error while processing payload: {}", p, t);
+        } finally {
+            if (!released) {
+                p.release();
+            }
+        }
+    }
+
+    @Override
+    public boolean mayTakeOwnership() {
+        return true;
+    }
+
     @Override
     public String getName() {
         return delegate.getName() + "-async";
     }
 
     @Override
     public boolean process(Payload payload) {
-        return payloads.offerFirst(payload);
+        boolean enqueued = payloads.offer(payload);
+        if (!enqueued) {
+            dropped.incrementAndGet();
+        }
+        return enqueued;
+    }
+
+    @Override
+    public void close() throws IOException {
+        this.delegate.close();
+        this.executorService.shutdown();
```

Comment:
Would be best to change this to
```java
    this.executorService.shutdownNow();
```
which should interrupt the running worker thread (which is what we want).

and move the `delegate.close()` to the end of the worker run method so it's called before the task exits.

Reply:
agreed, will do this.